File: src/test/java/org/railwaystations/rsapi/adapter/in/web/controller/CountriesControllerTest.java
package org.railwaystations.rsapi.adapter.in.web.controller;

import com.fasterxml.jackson.databind.ObjectMapper;
import org.jetbrains.annotations.NotNull;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.ValueSource;
import org.railwaystations.rsapi.adapter.in.web.ErrorHandlingControllerAdvice;
import org.railwaystations.rsapi.adapter.in.web.model.CountryDto;
import org.railwaystations.rsapi.adapter.out.db.CountryDao;
import org.railwaystations.rsapi.core.model.Country;
import org.railwaystations.rsapi.core.model.License;
import org.railwaystations.rsapi.core.model.ProviderApp;
import org.railwaystations.rsapi.core.services.CountryService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.test.context.ContextConfiguration;
import org.springframework.test.web.servlet.MockMvc;

import java.util.List;
import java.util.Set;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Fail.fail;
import static org.mockito.Mockito.when;
import static org.railwaystations.rsapi.utils.OpenApiValidatorUtil.validOpenApiResponse;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

@WebMvcTest(controllers = CountriesController.class)
@ContextConfiguration(classes = {WebMvcTestApplication.class, ErrorHandlingControllerAdvice.class, CountryService.class})
@AutoConfigureMockMvc(addFilters = false)
class CountriesControllerTest {

    @Autowired
    private MockMvc mvc;

    @MockBean
    private CountryDao countryDao;

    @Autowired
    private ObjectMapper objectMapper;

    @ParameterizedTest
    @ValueSource(strings = {"/countries", "/countries.json"})
    void listCountries(String urlTemplate) throws Exception {
        when(countryDao.list(true)).thenReturn(createCountryList());

        var contentAsString = mvc.perform(get(urlTemplate))
                .andExpect(status().isOk())
                .andExpect(validOpenApiResponse())
                .andReturn().getResponse().getContentAsString();

        List<CountryDto> countries = objectMapper.readerForListOf(CountryDto.class).readValue(contentAsString);
        assertThat(countries.size()).isEqualTo(2);
        countries.forEach(this::assertCountry);
    }

    @NotNull
    private Set<Country> createCountryList() {
        return Set.of(createCountry("xy"), createCountry("ab"));
    }

    @NotNull
    private Country createCountry(String code) {
        var country = Country.builder()
                .code(code)
                .name("name-" + code)
                .email("email-" + code)
                .twitterTags("twitter-" + code)
                .timetableUrlTemplate("timetable-" + code)
                .overrideLicense(License.CC_BY_NC_40_INT)
                .active(true)
                .build();
        country.getProviderApps().add(createProviderApp("android", code));
        country.getProviderApps().add(createProviderApp("ios", code));
        country.getProviderApps().add(createProviderApp("web", code));
        return country;
    }

    private ProviderApp createProviderApp(String type, String code) {
        return ProviderApp.builder().type(type).name("Provider-" + code).url(type + "App-" + code).build();
    }

    private void assertCountry(CountryDto country) {
        assertThat(country.getName()).isEqualTo("name-" + country.getCode());
        assertThat(country.getEmail()).isEqualTo("email-" + country.getCode());
        assertThat(country.getTwitterTags()).isEqualTo("twitter-" + country.getCode());
        assertThat(country.getTimetableUrlTemplate()).isEqualTo("timetable-" + country.getCode());
        assertThat(country.getOverrideLicense()).isEqualTo(License.CC_BY_NC_40_INT.getDisplayName());
        assertThat(country.getProviderApps().size()).isEqualTo(3);
        country.getProviderApps().forEach(app -> {
            switch (app.getType()) {
                case ANDROID -> assertThat(app.getUrl()).isEqualTo("androidApp-" + country.getCode());
                case IOS -> assertThat(app.getUrl()).isEqualTo("iosApp-" + country.getCode());
                case WEB -> assertThat(app.getUrl()).isEqualTo("webApp-" + country.getCode());
                default -> fail("unknown app type");
            }
        });
    }

}

File: adapter/src/test/kotlin/org/railwaystations/rsapi/adapter/web/controller/CountriesControllerTest.kt
package org.railwaystations.rsapi.adapter.web.controller

import com.fasterxml.jackson.databind.ObjectMapper
import com.ninjasquad.springmockk.MockkBean
import io.mockk.every
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Fail
import org.junit.jupiter.api.Test
import org.railwaystations.rsapi.adapter.db.CountryDao
import org.railwaystations.rsapi.adapter.web.ErrorHandlingControllerAdvice
import org.railwaystations.rsapi.adapter.web.OpenApiValidatorUtil.validOpenApiResponse
import org.railwaystations.rsapi.adapter.web.model.CountryDto
import org.railwaystations.rsapi.adapter.web.model.ProviderAppDto
import org.railwaystations.rsapi.core.model.CountryTestFixtures.createCountryList
import org.railwaystations.rsapi.core.model.License
import org.railwaystations.rsapi.core.services.CountryService
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc
import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest
import org.springframework.test.context.ContextConfiguration
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get
import org.springframework.test.web.servlet.result.MockMvcResultMatchers.status
import java.util.function.Consumer

@WebMvcTest(controllers = [CountriesController::class])
@ContextConfiguration(classes = [WebMvcTestApplication::class, ErrorHandlingControllerAdvice::class, CountryService::class])
@AutoConfigureMockMvc(addFilters = false)
internal class CountriesControllerTest {

    @Autowired
    private lateinit var mvc: MockMvc

    @Autowired
    private lateinit var objectMapper: ObjectMapper

    @MockkBean
    private lateinit var countryDao: CountryDao

    @Test
    fun listCountries() {
        every { countryDao.list(true) } returns createCountryList(setOf("xy", "ab"))

        val contentAsString = mvc.perform(get("/countries"))
            .andExpect(status().isOk())
            .andExpect(validOpenApiResponse())
            .andReturn().response.contentAsString

        val countries =
            objectMapper.readerForListOf(CountryDto::class.java).readValue<List<CountryDto>>(contentAsString)
        assertThat(countries.size).isEqualTo(2)
        countries.forEach(Consumer { country -> assertCountry(country) })
    }

}

fun assertCountry(country: CountryDto) {
    assertThat(country.name).isEqualTo("name-" + country.code)
    assertThat(country.email).isEqualTo("email-" + country.code)
    assertThat(country.timetableUrlTemplate).isEqualTo("timetable-" + country.code)
    assertThat(country.overrideLicense).isEqualTo(License.CC_BY_NC_40_INT.displayName)
    assertThat(country.providerApps!!.size).isEqualTo(3)
    country.providerApps!!.forEach(Consumer { app ->
        when (app.type) {
            ProviderAppDto.Type.ANDROID -> assertThat(app.url).isEqualTo("androidApp-" + country.code)
            ProviderAppDto.Type.IOS -> assertThat(app.url).isEqualTo("iosApp-" + country.code)
            ProviderAppDto.Type.WEB -> assertThat(app.url).isEqualTo("webApp-" + country.code)
            else -> Fail.fail<Any>("unknown app type")
        }
    })
}
